File: Problems/BinaryTreeVerticalOrderTraversal.cpp
#include <iostream>
#include <vector>
#include <map>
#include <queue>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    vector<vector<int>> verticalOrder(TreeNode *root)
    {
        map<int, vector<int>> map;
        queue<pair<int, TreeNode *>> queue;
        queue.push({0, root});
        while (!queue.empty())
        {
            pair<int, TreeNode *> p = queue.front();
            queue.pop();
            if (p.second != NULL)
            {
                if (map.find(p.first) == map.end())
                    map.insert(pair<int, vector<int>>(p.first, vector<int>()));
                map[p.first].push_back(p.second->val);
                queue.push({p.first - 1, p.second->left});
                queue.push({p.first + 1, p.second->right});
            }
        }

        vector<vector<int>> res;
        for (const auto &s : map)
            res.push_back(s.second);
        return res;
    }
};

File: Problems/BinaryTreeVerticalOrderTraversal.py
from collections import OrderedDict, defaultdict, deque
from typing import List, Optional


class TreeNode:
    def __init__(self, val, left, right):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def verticalOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if root is None:
            return []
        dic = {}
        dq = deque()
        dq.append((root, 50))
        while len(dq) > 0:
            node = dq.popleft()
            if node[1] not in dic:
                dic[node[1]] = [node[0].val]
            else:
                dic[node[1]].append(node[0].val)
            if node[0].left is not None:
                dq.append((node[0].left, node[1] - 1))
            if node[0].right is not None:
                dq.append((node[0].right, node[1] + 1))
        return [x for x in OrderedDict(sorted(dic.items())).values()]

    def verticalOrder2(self, root: TreeNode) -> List[List[int]]:
        columnTable = defaultdict(list)
        queue = deque([(root, 0)])

        while queue:
            node, column = queue.popleft()

            if node is not None:
                columnTable[column].append(node.val)

                queue.append((node.left, column - 1))
                queue.append((node.right, column + 1))

        return [columnTable[x] for x in sorted(columnTable.keys())]


e = TreeNode(7, None, None)
d = TreeNode(15, None, None)
c = TreeNode(20, d, e)
b = TreeNode(9, None, None)
a = TreeNode(3, b, c)
sol = Solution()
print(sol.verticalOrder(a))
